README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# acroname

[![CRAN
Status](https://www.r-pkg.org/badges/version/acroname)](https://cran.r-project.org/package=acroname)
![](https://cranlogs.r-pkg.org/badges/acroname)

## Installation

``` r
devtools::install_github("vpnagraj/acroname")
```

## Features

  - `acronym()`: Generate acronym from input string
  - `initialism()`: Generate initialism from input string
  - `mince()`: Prepare input string

## Usage

### Basic example

The minimal usage for creating an acronym:

``` r
library(acroname)
set.seed(123)
acronym("super special software package", acronym_length = 5)
#> [1] "SUSIE: SUper SpecIal softwarE package"
```

And an initialism for the same string:

``` r
initialism("super special software package")
#> [1] "SSSP: Super Special Software Package"
```

The functions can accept input character vectors that either contain the
string in a single element or are prepared with one element per word:

``` r
acronym(c("super", "special", "software", "package"), acronym_length = 5)
#> [1] "SUSIE: SUper SpecIal softwarE package"
```

### Convert result to `tibble`

Both `acronym()` and `initialism()` include an option to return the
result as a `tibble` object:

``` r
acronym("dip him in the river who loves water", acronym_length = 4, to_tibble = TRUE)
#> # A tibble: 1 x 4
#>   formatted                  prefix suffix                 original             
#>   <chr>                      <chr>  <chr>                  <chr>                
#> 1 ROOT: dip him in River wh… ROOT   dip him in River whO … dip him in river who…
```

### Exclude articles

Each function can also be customized to exclude articles in the input
string:

``` r
initialism("dip him in the river who loves water", to_tibble = TRUE, ignore_articles = TRUE)
#> # A tibble: 1 x 4
#>   formatted                   prefix  suffix               original             
#>   <chr>                       <chr>   <chr>                <chr>                
#> 1 DHIRWLW: Dip Him In River … DHIRWLW Dip Him In River Wh… dip him in river who…
```

### Trigger “bag of words” approach

Using the “bow” option will trigger processing using a “bag of words”
method, by which words in the input string are randomly selected. The
number of words selected from the input string depends on the value
passed to the “bow\_prop” argument:

``` r
acronym("dip him in the river who loves water the fountain contains the cistern overflows", ignore_articles = TRUE, acronym_length = 4, bow = TRUE, bow_prop = 0.75)
#> [1] "INRI: dip loves hIm who water fouNtain River In cistern"
```

It is possible to generate a series of randomized results by iterating
over the function:

``` r
library(purrr)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

## write a function to wrap acronym() in order to map
mapwrap <- function(i, input) {
  res <- 
    acronym(input = input, ignore_articles = TRUE, acronym_length =  4, bow = TRUE, bow_prop = 0.75, to_tibble = TRUE) %>%
    mutate(iteration = i, .before = "formatted")
  return(res)
}

iterations <- paste0("iteration_",1:5)
map_df(iterations, mapwrap, "dip him in the river who loves water the fountain contains the cistern overflows")
#> # A tibble: 5 x 5
#>   iteration  formatted              prefix suffix            original           
#>   <chr>      <chr>                  <chr>  <chr>             <chr>              
#> 1 iteration… CORA: dip Cistern whO… CORA   dip Cistern whO … dip cistern who ri…
#> 2 iteration… HOWE: Him who cistern… HOWE   Him who cistern … him who cistern ov…
#> 3 iteration… CROW: loves dip who C… CROW   loves dip who Co… loves dip who cont…
#> 4 iteration… INST: IN him containS… INST   IN him containS … in him contains ci…
#> 5 iteration… DIOR: cistern who DIp… DIOR   cistern who DIp … cistern who dip ov…
```

### Use a custom dictionary

By default the `acronym()` engine will search for acronyms that match
words in the “en\_US” dictionary provided by the
[hunspell](https://CRAN.R-project.org/package=hunspell) package.
However, the dictionary of words to match can be customized via the
“dictionary” argument. The example below uses a dictionary based on
names from the dplyr package, prepared with a little help from the
`mince()` function:

``` r
## get names (first or last) from the dplyr::starwars data
## see mince() in action
mince(starwars$name)
#> $words
#>   [1] "Luke"        "Skywalker"   "C3PO"        "R2D2"        "Darth"      
#>   [6] "Vader"       "Leia"        "Organa"      "Owen"        "Lars"       
#>  [11] "Beru"        "Whitesun"    "lars"        "R5D4"        "Biggs"      
#>  [16] "Darklighter" "ObiWan"      "Kenobi"      "Anakin"      "Skywalker"  
#>  [21] "Wilhuff"     "Tarkin"      "Chewbacca"   "Han"         "Solo"       
#>  [26] "Greedo"      "Jabba"       "Desilijic"   "Tiure"       "Wedge"      
#>  [31] "Antilles"    "Jek"         "Tono"        "Porkins"     "Yoda"       
#>  [36] "Palpatine"   "Boba"        "Fett"        "IG88"        "Bossk"      
#>  [41] "Lando"       "Calrissian"  "Lobot"       "Ackbar"      "Mon"        
#>  [46] "Mothma"      "Arvel"       "Crynyd"      "Wicket"      "Systri"     
#>  [51] "Warrick"     "Nien"        "Nunb"        "QuiGon"      "Jinn"       
#>  [56] "Nute"        "Gunray"      "Finis"       "Valorum"     "Jar"        
#>  [61] "Jar"         "Binks"       "Roos"        "Tarpals"     "Rugor"      
#>  [66] "Nass"        "Ric"         "Olié"        "Watto"       "Sebulba"    
#>  [71] "Quarsh"      "Panaka"      "Shmi"        "Skywalker"   "Darth"      
#>  [76] "Maul"        "Bib"         "Fortuna"     "Ayla"        "Secura"     
#>  [81] "Dud"         "Bolt"        "Gasgano"     "Ben"         "Quadinaros" 
#>  [86] "Mace"        "Windu"       "KiAdiMundi"  "Kit"         "Fisto"      
#>  [91] "Eeth"        "Koth"        "Adi"         "Gallia"      "Saesee"     
#>  [96] "Tiin"        "Yarael"      "Poof"        "Plo"         "Koon"       
#> [101] "Mas"         "Amedda"      "Gregar"      "Typho"       "Cordé"      
#> [106] "Cliegg"      "Lars"        "Poggle"      "Lesser"      "Luminara"   
#> [111] "Unduli"      "Barriss"     "Offee"       "Dormé"       "Dooku"      
#> [116] "Bail"        "Prestor"     "Organa"      "Jango"       "Fett"       
#> [121] "Zam"         "Wesell"      "Dexter"      "Jettster"    "Lama"       
#> [126] "Su"          "Taun"        "We"          "Jocasta"     "Nu"         
#> [131] "Ratts"       "Tyerell"     "R4P17"       "Wat"         "Tambor"     
#> [136] "San"         "Hill"        "Shaak"       "Ti"          "Grievous"   
#> [141] "Tarfful"     "Raymus"      "Antilles"    "Sly"         "Moore"      
#> [146] "Tion"        "Medon"       "Finn"        "Rey"         "Poe"        
#> [151] "Dameron"     "BB8"         "Captain"     "Phasma"      "Padmé"      
#> [156] "Amidala"    
#> 
#> $collapsed
#> [1] "LukeSkywalkerC3POR2D2DarthVaderLeiaOrganaOwenLarsBeruWhitesunlarsR5D4BiggsDarklighterObiWanKenobiAnakinSkywalkerWilhuffTarkinChewbaccaHanSoloGreedoJabbaDesilijicTiureWedgeAntillesJekTonoPorkinsYodaPalpatineBobaFettIG88BosskLandoCalrissianLobotAckbarMonMothmaArvelCrynydWicketSystriWarrickNienNunbQuiGonJinnNuteGunrayFinisValorumJarJarBinksRoosTarpalsRugorNassRicOliéWattoSebulbaQuarshPanakaShmiSkywalkerDarthMaulBibFortunaAylaSecuraDudBoltGasganoBenQuadinarosMaceWinduKiAdiMundiKitFistoEethKothAdiGalliaSaeseeTiinYaraelPoofPloKoonMasAmeddaGregarTyphoCordéClieggLarsPoggleLesserLuminaraUnduliBarrissOffeeDorméDookuBailPrestorOrganaJangoFettZamWesellDexterJettsterLamaSuTaunWeJocastaNuRattsTyerellR4P17WatTamborSanHillShaakTiGrievousTarffulRaymusAntillesSlyMooreTionMedonFinnReyPoeDameronBB8CaptainPhasmaPadméAmidala"
#> 
#> $words_len
#>   [1]  4  9  4  4  5  5  4  6  4  4  4  8  4  4  5 11  6  6  6  9  7  6  9  3  4
#>  [26]  6  5  9  5  5  8  3  4  7  4  9  4  4  4  5  5 10  5  6  3  6  5  6  6  6
#>  [51]  7  4  4  6  4  4  6  5  7  3  3  5  4  7  5  4  3  4  5  7  6  6  4  9  5
#>  [76]  4  3  7  4  6  3  4  7  3 10  4  5 10  3  5  4  4  3  6  6  4  6  4  3  4
#> [101]  3  6  6  5  5  6  4  6  6  8  6  7  5  5  5  4  7  6  5  4  3  6  6  8  4
#> [126]  2  4  2  7  2  5  7  5  3  6  3  4  5  2  8  7  6  8  3  5  4  5  4  3  3
#> [151]  7  3  7  6  5  7
#> 
#> $first_chars
#>   [1] "L" "S" "C" "R" "D" "V" "L" "O" "O" "L" "B" "W" "l" "R" "B" "D" "O" "K"
#>  [19] "A" "S" "W" "T" "C" "H" "S" "G" "J" "D" "T" "W" "A" "J" "T" "P" "Y" "P"
#>  [37] "B" "F" "I" "B" "L" "C" "L" "A" "M" "M" "A" "C" "W" "S" "W" "N" "N" "Q"
#>  [55] "J" "N" "G" "F" "V" "J" "J" "B" "R" "T" "R" "N" "R" "O" "W" "S" "Q" "P"
#>  [73] "S" "S" "D" "M" "B" "F" "A" "S" "D" "B" "G" "B" "Q" "M" "W" "K" "K" "F"
#>  [91] "E" "K" "A" "G" "S" "T" "Y" "P" "P" "K" "M" "A" "G" "T" "C" "C" "L" "P"
#> [109] "L" "L" "U" "B" "O" "D" "D" "B" "P" "O" "J" "F" "Z" "W" "D" "J" "L" "S"
#> [127] "T" "W" "J" "N" "R" "T" "R" "W" "T" "S" "H" "S" "T" "G" "T" "R" "A" "S"
#> [145] "M" "T" "M" "F" "R" "P" "D" "B" "C" "P" "P" "A"

sw_names <- mince(starwars$name)$words

acronym("latest and greatest data analysis package", dictionary = sw_names, acronym_length = 4)
#> [1] "LARS: Latest And gReatest data analySis package"
```

### Configure timeout option

Depending on the characters in the input string, the size of dictionary,
and the acronym length it may not be possible to generate an acronym
that matches a word in the dictionary. The `acronym()` function will try
for 60 seconds by default, but this duration can be customized via the
“timeout” argument:

``` r
## the longer the desired acronym, the longer it will likely to take to find a match
## setting the timeout to 2 seconds here
acronym("latest and greatest data analysis package", dictionary = sw_names, acronym_length = 10, timeout = 2)
#> Unable to find viable acronym in 'timeout' specified (2 seconds) ...
```

## Related

  - <https://github.com/bacook17/acronym>

## Contributing

Please use GitHub issues to report bugs or request features.
Contributions will be reviewed via pull requests.


NEWS.md
# acroname 0.1.0

* Initial release!


tests/testthat.R
library(testthat)
library(acroname)

test_check("acroname")


tests/testthat/test-utils.R
test_that("first_char can extract first letter or number", {

  tst <- first_char("NOTHING")
  expect_equal(tst, "N")

  tst <- first_char("compares")
  expect_equal(tst, "c")

  tst <- first_char("2U")
  expect_equal(tst, "2")

})

test_that("find_articles works as expected", {

  tst <- find_articles("a")
  expect_true(tst)

  tst <- find_articles("AN")
  expect_true(tst)

  tst <- find_articles("The")
  expect_true(tst)

  tst <- find_articles("whatever")
  expect_true(!tst)

})

test_that("mince works as expected with defaults", {

  tst <- mince("the fox condemns the trap, not himself")

  expect_equal(tst$words, c("fox", "condemns", "trap", "not", "himself"))
  expect_equal(tst$collapsed, "foxcondemnstrapnothimself")
  expect_equal(tst$words_len, c(3,8,4,3,7))
  expect_equal(tst$first_chars, c("f","c","t","n","h"))

})

test_that("mince works with articles", {

  tst <- mince("the fox condemns the trap, not himself", ignore_articles = FALSE)

  expect_equal(tst$words, c("the","fox", "condemns", "the", "trap", "not", "himself"))
  expect_equal(tst$collapsed, "thefoxcondemnsthetrapnothimself")
  expect_equal(tst$words_len, c(3,3,8,3,4,3,7))
  expect_equal(tst$first_chars, c("t","f","c","t","t","n","h"))

})


test_that("mince works with non-alphanumeric", {

  tst <- mince("the fox condemns the trap, not himself", alnum_only = FALSE, ignore_articles = FALSE)

  expect_equal(tst$words, c("the","fox", "condemns", "the", "trap,", "not", "himself"))
  expect_equal(tst$collapsed, "thefoxcondemnsthetrap,nothimself")
  expect_equal(tst$words_len, c(3,3,8,3,5,3,7))
  expect_equal(tst$first_chars, c("t","f","c","t","t","n","h"))

})


tests/testthat/test-engines.R
test_that("inititalism captures first chars", {

  tst <- initialism("the fox condemns the trap, not himself")
  expect_equal(tst, "FCTNH: Fox Condemns Trap Not Himself")

  tst <- initialism("the fox condemns the trap, not himself", ignore_articles = FALSE)
  expect_equal(tst, "TFCTTNH: The Fox Condemns The Trap Not Himself")

  tst <- initialism("Nothing compares 2 u")
  expect_equal(tst, "NC2U: Nothing Compares 2 U")

})


R/utils.R
#' Extract the first character from a string
#'
#' @description
#'
#' This helper function will extract the first character from a string. The element may be a letter, number, or special character but will be coerced to a character vector in the output.
#'
#' @param string Character vector from which the first character will be extracted
#'
#' @return Character vector with the first character from each element in the vector passed to the input "string" argument. This value will be the same length as the original vector.
#'
#' @export
#'
#' @md
#'
#' @examples
#'
#' first_char("purple")
#' first_char(c("purple","rain"))
#' first_char(c("nothing","compares","2u"))
#'
first_char <- function(string) {
  string %>%
    stringr::str_split(., pattern = "") %>%
    purrr::map_chr(., .f = function(x) utils::head(x, 1))
}


#' Helper to find articles
#'
#' @description
#'
#' This function will check if an input word is an article in the English language ('a', 'an', 'the').
#'
#' @param word Character vector of length 1 with word to check
#'
#' @return Logical vector of length one, `TRUE` if the word is an article and `FALSE` if not.
#' @md
#' @export
#'
#' @examples
#'
#' find_articles("the")
#' find_articles("then")
#' find_articles("whatever")
#'
find_articles <- function(word) {
  word <- tolower(word)
  any(c("a","an","the") %in% word)
}

#' Find candidate
#'
#' @description
#'
#' This is an unexported helper for \link[acroname]{acronym}. The function is used wrapped in a `tryCatch()` that uses \link[R.utils]{withTimeout} to manage maximum wait time for the candidate acronym search.
#'
#' @param collapsed The collapsed string of characters generated by \link[acroname]{mince}
#' @param acronym_length Number of characters in acronym; default is `3`
#' @param probs Vector of probabilities for selecting each character while generating candidate
#' @param dictionary Character vector containing dictionary of terms from which acronym should be created; default is `NULL` and `hunspell` "en_us" dictionary will be used
#' @param words_len Vector of the length of each word in the input
#'
#' @md
#' @return Named list with three elements:
#'
#' - **formatted**: The candidate acronym and string with letters used capitalized
#' - **prefix**: The candidate acronym
#' - **suffix**: Words used with letters in acronym capitalized
#'#'
find_candidate <- function(collapsed, acronym_length, probs, dictionary, words_len) {
  valid <- FALSE
  while(!valid) {
    ## sample the *indices* of possible letters to use
    ## use the length of the acronym for the size out
    ## use probability weighting
    indices <- sort(sample(1:nchar(collapsed), size = acronym_length, replace = FALSE, prob = probs))

    tmp_collapsed_split <-
      collapsed %>%
      strsplit(., split = "") %>%
      unlist(.)

    candidate <- paste0(tmp_collapsed_split[indices], collapse = "")

    if(candidate %in% dictionary) {
      valid <- TRUE
      ## now format the output to include the capitalized letter
      tmp_collapsed_split <- tolower(tmp_collapsed_split)
      tmp_collapsed_split[indices] <- toupper(tmp_collapsed_split[indices])
      ## now need to split the words up again
      last_letter_ind <- cumsum(words_len)
      tmp_collapsed_split[last_letter_ind[-length(last_letter_ind)]] <- paste0(tmp_collapsed_split[last_letter_ind[-length(last_letter_ind)]], " ")
      name <-  paste0(tmp_collapsed_split, collapse = "")
    } else {
      valid <- FALSE
      candidate <- ""
      name <- ""
    }
  }

  ## format with original
  formatted <- paste0(toupper(candidate), ": ", name)

  res <-
    list(
      formatted = formatted,
      prefix = toupper(candidate),
      suffix = name
    )

  return(res)
}


#' Prepare input string
#'
#' @description
#'
#' This helper is used by both \link[acroname]{acronym} and \link{initialism} to extract elements needed from the input string.
#'
#' If the function is used with `bow = TRUE` the input will be processed with a "bag of words" approach, by which words will be shuffled and sampled without replacement. In this case, the number of characters used will be determined by the proportion passed to "bow_prop".
#'
#' @param input Character vector with text to use as the input for the candidate
#' @param ignore_articles Logical indicating whether or not articles should be ignored ; default is `TRUE`
#' @param alnum_only Logical that specifes whether only alphanumeric should be used; default is `TRUE`
#' @param bow Logical for whether or not a "bag of words" approach should be used for "input" vector; default is `FALSE`
#' @param bow_prop Given `bow = TRUE` this specifies the proportion of words to sample; ignored if `bow = FALSE`; default is `0.5`
#'
#' @return Named list with the following elements:
#'
#' - **words**: Vector with one element per word to be used in the acronym or initialism
#' - **collapsed**: Vector of length 1 containing all characters from words collapsed
#' - **words_len**: Vector containing length of each word
#' - **first_chars**: Vector containing first character from each word
#'
#' @md
#' @export
#'
mince <- function(input, ignore_articles = TRUE, alnum_only = TRUE, bow = FALSE, bow_prop = 0.5) {

  ## check if the input character vector is > 1
  ## if so ... collapse together prior to splitting
  ## this allows input to be either "purple rain" or c("purple","rain")
  if(length(input) > 1) {
    input <- paste0(input, collapse = " ")
  }

  ## get the words split into a vector with one element per word
  words <-
    strsplit(input, " ") %>%
    unlist(.)

  ## handle alphanumeric only option
  ## replace any non-numeric with ""
  if(alnum_only) {
    words <-
      words %>%
      stringr::str_replace_all(., "[^[:alnum:]]", "")
  }

  ## use this vector of words to conditionally check for articles
  if(ignore_articles) {
    ## find the indices for words that are articles
    articles <-
      words %>%
      purrr::map_lgl(., .f = find_articles)

    ## find the indices for words that are *not* articles
    not_articles <- which(!articles)
    ## and keep them
    words <- words[not_articles]
  }

  ## handle the bag-of-words option
  ## if
  if(bow) {
    words <- sample(words, size = ceiling(bow_prop*length(words)), replace = FALSE)
  }

  ## find the number of letters in each word
  char_len <-
    words %>%
    purrr::map_dbl(., nchar)

  ## find the first characters of each word
  first_chars <-
    words %>%
    purrr::map_chr(., first_char)

  ## collapse words a vector with one element and all letters
  collapsed <-
    words %>%
    paste0(., collapse = "")

  ## return a named list with relevant prepped data
  list(
    words = words,
    collapsed = collapsed,
    words_len = char_len,
    first_chars = first_chars
  )

}


R/engines.R
#' acroname engines
#' @name engines
#'
#' @description
#'
#' The `acroname` engines include methods to generate acronyms and initialisms. `acronym()` searches for candidates by constructing words from characters provided. Each word constructed is compared to the terms in the dictionary specified, and once a match is found the acronym is returned. `initialism()` takes the first characters from each word in the string. Both functions can optionally return a `tibble`, ignore articles, and/or use a "bag of words" approach (for more see \link[acroname]{mince}).
#'
#' @param input Character vector with text to use as the input for the candidate
#' @param dictionary Character vector containing dictionary of terms from which acronym should be created; default is `NULL` and `hunspell` "en_us" dictionary will be used
#' @param acronym_length Number of characters in acronym; default is `3`
#' @param ignore_articles Logical indicating whether or not articles should be ignored ; default is `TRUE`
#' @param alnum_only Logical that specifes whether only alphanumeric should be used; default is `TRUE`
#' @param timeout Maximum seconds to spend searching for an acronym; default is `60`
#' @param bow Logical for whether or not a "bag of words" approach should be used for "input" vector; default is `FALSE`
#' @param bow_prop Given `bow = TRUE` this specifies the proportion of words to sample; ignored if `bow = FALSE`; default is `0.5`
#' @param to_tibble Logical as to whether or not the result should be a `tibble`; default is `FALSE`
#'
#' @return
#'
#' If `to_tibble = FALSE` (default), then a character vector containing the name capitalized followed by the original string with letters used in the name capitalized.
#'
#' If `to_tibble = TRUE`, then a `tibble` with the following columns:
#'
#' - **formatted**: The candidate name and string with letters used capitalized
#' - **prefix**: The candidate name
#' - **suffix**: Words used with letters in name capitalized
#' - **original**: The original string used to construct the name
#'
#' @md
#'

#' @export
#' @rdname engines
acronym <- function(input, dictionary = NULL, acronym_length = 3, ignore_articles = TRUE, alnum_only = TRUE, timeout = 60, bow = FALSE, bow_prop = 0.5, to_tibble = FALSE) {

  ## default behavior is to use hunspell en_us
  if(is.null(dictionary)) {

    dictpath <- system.file("dict", "en_US.dic", package = "hunspell")
    stopifnot(file.exists(dictpath))

    dictionary <-
      dictpath %>%
      readr::read_lines() %>%
      grep("^[A-Z]", ., value = TRUE) %>%
      gsub("/[A-Z]*.", "", .) %>%
      tolower()
  }

  tmp <- mince(input = input,
               ignore_articles = ignore_articles,
               alnum_only = alnum_only,
               bow = bow,
               bow_prop = bow_prop)

  ## get the index of the first character
  first_char_ind <- abs(tmp$words_len  - cumsum(tmp$words_len))  + 1

  ## inintially set all indicies low
  probs <- rep(0.1, nchar(tmp$collapsed))
  ## for first characters bump up probability that index will be selected much higher
  probs[first_char_ind] <- 0.9
  ## bump up weight of first letter ?!
  probs[1] <- 0.95

  ## use timeout with tryCatch
  ## if the processing takes longer than timeout then a message will be printed
  tryCatch({
     R.utils::withTimeout({
      res <- find_candidate(collapsed = tmp$collapsed,
                            acronym_length = acronym_length,
                            probs = probs,
                            dictionary = dictionary,
                            words_len = tmp$words_len)

      if(to_tibble) {
        ## tibble from result elements
        res_tibble <-
          dplyr::tibble(
            formatted = res$formatted,
            prefix = res$prefix,
            suffix = res$suffix,
            original = paste0(tmp$words, collapse = " ")
          )
        return(res_tibble)
      } else {
        return(res$formatted)
      }

    }, timeout = timeout)
  }, TimeoutException = function(ex) {
    message(sprintf("Unable to find viable acronym in 'timeout' specified (%d seconds) ... ", timeout))
  })

}


#' @export
#' @rdname engines
initialism <- function(input, ignore_articles = TRUE, alnum_only = TRUE, bow = FALSE, bow_prop = 0.5, to_tibble = FALSE) {

  ## process input
  tmp <- mince(input = input,
               ignore_articles = ignore_articles,
               alnum_only = alnum_only,
               bow = bow,
               bow_prop = bow_prop)

  ## get candidate prefix
  candidate <- paste0(toupper(tmp$first_chars), collapse = "")

  tmp_collapsed_split <-
    tmp$collapsed %>%
    strsplit(., split = "") %>%
    unlist(.)

  ## indices for first char
  ## get the index of the first character
  first_char_ind <- abs(tmp$words_len  - cumsum(tmp$words_len))  + 1

  ## now format the output to include the capitalized letter
  ## first force everything to be lower case
  tmp_collapsed_split <- tolower(tmp_collapsed_split)
  ## then make sure letters at first char indices are capitalized
  tmp_collapsed_split[first_char_ind] <- toupper(tmp_collapsed_split[first_char_ind])
  ## now need to split the words up again
  last_letter_ind <- cumsum(tmp$words_len)
  tmp_collapsed_split[last_letter_ind[-length(last_letter_ind)]] <- paste0(tmp_collapsed_split[last_letter_ind[-length(last_letter_ind)]], " ")
  name  <- paste0(tmp_collapsed_split, collapse = "")

  ## format with original
  formatted <- paste0(toupper(candidate), ": ", name)

  if(to_tibble) {
    ## tibble from result elements
    res_tibble <-
      dplyr::tibble(
        formatted = formatted,
        prefix = candidate,
        suffix = name,
        original = paste0(tmp$words, collapse = " ")
      )
    return(res_tibble)
  } else {
    return(formatted)
  }


}
